modules/p2/ding_interests/ding_interests.module

```diff
@@ -0,0 +1,652 @@
+  $list = ding_list_get_list(DING_INTERESTS_LIST);
+  if (empty($list)) {
+    return array();
+  }
+  $elements = ding_list_load_elements($list);
+  $interests = array();
+  foreach ($elements as $element) {
+    $element_id = ding_list_get_list_element_value($element);
+    if (!$type || ding_list_get_list_element_state($element) == $type) {
+      $interests[] = $element_id;
+    }
+  }
+  return $interests;
+}
+
+/**
+ * Get random interest element.
+ *
+ * @return object
+ *   Taxonomy term corresponding to the interest.
+ */
+function ding_interests_get_random_interest_term() {
+  $term = ding_interests_get_random_list_item();
+  return $term;
+}
+
+/**
+ * Implements hook_user_complete().
+ */
+function ding_interests_ding_user_complete() {
+  $complete = FALSE;
+  return array(
+    'ding_interests' => array(
+      'title' => t('Follow topics from the library'),
+      'points' => 3,
+      'completed' => $complete,
+      'markup' =>
+      "<div class='description'>" .
+      t('You could start building a list of your interests to make your profile more complete.') .
+      "</div><ul class='actions'><li>" .
+      l(t("Check out the topics"), "ding/interests/overview") .
+      "</li></ul>",
+    ),
+  );
+}
+
+/**
+ * Fetch some random ding_object from user list.
+ *
+ * @return object
+ *   The random ting_object. Or FALSE otherwise.
+ */
+function ding_interests_get_random_list_item() {
+  $list = ding_list_get_list(DING_INTERESTS_LIST);
+  if (empty($list)) {
+    return FALSE;
+  }
+
+  // @TODO: add criteria to ding_list_load_elements so we dont have to load all
+  // @TODO: add get random elements to ding_list
+  $items = ding_list_load_elements($list);
+  $buffer = array();
+  foreach ($items as $item) {
+    if (ding_base_get_value('ding_type', $item, 'field_state') == 'yes') {
+      $buffer[] = $item;
+    }
+  }
+
+  if (empty($buffer)) {
+    return FALSE;
+  }
+
+  $rnd = array_rand($buffer);
+  $element = $buffer[$rnd];
+
+  $term_name = ding_base_get_value('ding_type', $element, 'field_search_title');
+  $terms = taxonomy_get_term_by_name($term_name, 'ding_content_tags');
+
+  // Missing term.
+  if (count($terms) == 0) {
+    return FALSE;
+  }
+
+  $access = ding_interests_get_status_tids();
+  foreach ($terms as $term) {
+    if ($term->field_status[LANGUAGE_NONE][0]['tid'] == $access['granted']) {
+      return $term;
+    }
+  }
+
+  // Return any.
+  return array_values($terms)[0];
+}
+
+
+/**
+ * Implements hook_theme().
+ */
+function ding_interests_theme($existing, $type, $theme, $path) {
+  $theme = array(
+    'ding_interests_tree' => array(
+      // 'template' => 'ding-in',
+      'variables' => array('root' => 0, 'tree' => array()),
+      // 'path' => $path . '/theme',
+    ),
+  );
+  return $theme;
+}
+
+/**
+ * Preprocess function for Ding interests: Overview.
+ */
+function template_preprocess_ding_interests_tree(&$variables) {
+  $voc = taxonomy_vocabulary_machine_name_load('ding_content_tags');
```

Comment:
What is the purpose of all this preprocessing? There seems to be a lot going and some code comments would help explain why.